mdm/common/src/main/java/com/gentics/mesh/core/data/dao/PersistingLanguageDao.java

```diff
@@ -23,6 +35,42 @@ default HibLanguage create(String languageName, String languageTag, String uuid)
 		return language;
 	}
 
+	@Override
+	default void assign(HibLanguage language, HibProject project, EventQueueBatch batch, boolean throwOnExisting) {
+		project.getLanguages().stream()
+			.filter(l -> l.getLanguageTag().equals(language.getLanguageTag()))
+			.findAny()
+			.ifPresentOrElse(existing -> {
+					if (throwOnExisting) {
+						throw error(BAD_REQUEST, "error_language_already_assigned", language.getLanguageTag(), project.getName());
+					} else {
+						log.debug("Language [{}] is already assigned to the project [{}]", language.getLanguageTag(), project.getName());
+					}
+				}, () -> {
+					project.addLanguage(language);
+				});
+	}
+
+	@Override
+	default void unassign(HibLanguage language, HibProject project, EventQueueBatch batch, boolean throwOnInexisting) {
+		project.getLanguages().stream()
+			.filter(l -> l.getLanguageTag().equals(language.getLanguageTag()))
+			.findAny()
+			.ifPresentOrElse(existing -> {
+					CommonTx.get().nodeDao().findUsedLanguages(project, Collections.singletonList(language.getLanguageTag()), true).stream().findAny().ifPresentOrElse(existingContent -> {
+						throw error(BAD_REQUEST, "error_language_still_in_use", language.getLanguageTag(), project.getName());
+					}, () -> {
+						project.removeLanguage(language);
+					});
+				}, () -> {
+					if (throwOnInexisting) {
+						throw error(BAD_REQUEST, "error_language_not_assigned", language.getLanguageTag(), project.getName());
```

Comment:
Isn't NOT_FOUND the better response code here?
Or should we remove this, because throwOnInexisting is currently set to false on all occasions?

Reply:
I'd rather leave both, since a) NOT_FOUND clearly stands for 'language not found for the context' (and there's hardly an alternative for BAD_REQUEST in such case, except maybe a CONFLICT), b) we don't know at the moment if it gonna be used with true in the future.